File: 19.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        if not head.next:
            return head if n == 0 else None
        if not head.next.next:
            if n == 0:
                return head
            elif n == 1:
                head.next = None
                return head
            else:
                return head
        sz = 0
        curr = head
        while curr:
            curr = curr.next
            sz += 1
        first, second = head, head.next
        sz -= 1
        while second.next and sz > 0:
            first = first.next
            second = second.next
            sz -= 1
            print(first.val)
            print(second.val)
            print("---")
        first.next = second.next
        return head

File: 19.cpp
#include <algorithm>
#include <deque>
#include <iostream>
#include <map>
#include <set>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
  void printList(ListNode *curr) {
    while (curr) {
      cout << curr->val << ", ";
      curr = curr->next;
    }
    cout << endl;
  }

public:
  ListNode *removeNthFromEnd(ListNode *head, int n) {
    ListNode *s = new ListNode(0, head);
    ListNode *f = head;
    while (f && n) {
      f = f->next;
      n -= 1;
    }
    if (!f) {
      return head->next;
    }
    while (f) {
      f = f->next;
      s = s->next;
    }
    s->next = s->next->next;
    return head;
  }
};
int main() {}
